testers/testers/markus_tester.py

```diff
@@ -379,38 +243,9 @@ def after_tester_run(self):
         """
         pass
 
+    @abstractmethod
     def run(self):
         """
         Runs this tester.
         """
-        try:
-            self.before_tester_run()
-            with contextlib.ExitStack() as stack:
-                feedback_open = (stack.enter_context(open(self.specs.feedback_file, 'w'))
-                                 if self.specs.feedback_file is not None
-                                 else None)
-                for test_file in sorted(self.specs.tests):
-                    test_extra = self.specs.matrix[test_file].get(MarkusTestSpecs.MATRIX_NONTEST_KEY, {})
-                    for data_files in sorted(self.specs.matrix[test_file].keys()):
-                        if data_files == MarkusTestSpecs.MATRIX_NONTEST_KEY:
-                            continue
-                        points = self.specs.matrix[test_file][data_files]
-                        if MarkusTestSpecs.DATA_FILES_SEPARATOR in data_files:
-                            data_files = data_files.split(MarkusTestSpecs.DATA_FILES_SEPARATOR)
-                        else:
-                            data_files = [data_files]
-                        test = self.test_class(self, test_file, data_files, points, test_extra, feedback_open)
-                        try:
-                            # if a test __init__ fails it should really stop the whole tester, we don't have enough
-                            # info to continue safely, e.g. the total points (which skews the student mark)
-                            self.before_test_run(test)
-                            xml = test.run()
-                            self.after_successful_test_run(test)
-                        except Exception as e:
-                            xml = test.error(message=str(e))
-                        finally:
-                            print(xml, flush=True)
-        except Exception as e:
-            print(MarkusTester.error_all(message=str(e)), flush=True)
-        finally:
-            self.after_tester_run()
+        pass
```

Comment:
Is there no longer a default way to run a tester?

Reply:
Not by running `MarkusTester` directly (it's now an abstract class). Every subclass of `MarkusTester` needs to define its own `run` method. The closest thing we have to a default now is to run the custom tester.

Reply:
That may be problematic:
- code duplication -> most testers will just have the same run method
- contract -> callbacks shouldn't be left to the testers to invoke, otherwise they're not callbacks anymore
Ideally, I think the tester loop should be as common as possible.

Reply:
If that is the case then we seriously need to rewrite most of the testers. Currently most of them do not have a common run method anyway.

I agree that we should make the testers more consistent but that should be in a future PR.

Reply:
sql and pyta did, java had a todo to address before being able to do so. The old python could not because of pam.
The new python, racket and haskell may have been able to use the default run method, but I didn't investigate.

All I'm saying is that all of the real custom behavior is isolated in the MarkusXTest.run method already, and we devised the callback mechanism exactly to avoid having to customize the MarkusXTester.run too, since most of the testers just do the same basic loop. If we have callbacks, we should have a default run, and put the rest into the callbacks.